lib/elixir/lib/macro.ex

```diff
@@ -245,11 +245,43 @@ defmodule Macro do
   end
 
   @doc """
-  Pipes `expr` into the `call_args` at the given `position`.
+  Pipes the `expr` into the `call_args` at the given `position`.
+
+  - `expr` Might be an AST, or just a literal value.
+  - `call_args` Must be an AST; otherwise error will be raised.
+
+  ## Examples
+
+      iex> ast = Macro.pipe(quote(do: div(10, 2)), quote(do: div(5)), 0)
+      {:div, [], [{:div, [context: MacroTest, import: Kernel], [10, 2]}, 5]}
+      iex> Code.eval_quoted(ast)
+      {1, []}
+
+  If attempted to pipe an expression into a non AST,
+  the following error will be raised:
+
+      iex> Macro.pipe(10, 20, 0)
+      ** (ArgumentError) cannot pipe 10 into 20, the 2nd argument must be an AST.
+
+  It's also good to remind that even if the
+  expression is piped into the AST, it doesn't mean that
+  the AST is valid, and it might generate runtime issues,
+  for example:
+
+      iex> ast = Macro.pipe(10, {:div, [], [2, 5]}, 0)
+      {:div, [], [10, 2, 5]}
+      iex> Code.eval_quoted(ast)
+      ** (CompileError) nofile:1: undefined function div/3 (there is no such import)
   """
   @spec pipe(t(), t(), integer) :: t()
   def pipe(expr, call_args, position)
 
+  def pipe(expr, args, _integer) when not is_tuple(args) do
+    raise ArgumentError,
+          "cannot pipe #{to_string(expr)} into #{to_string(args)}, " <>
+            "the 2nd argument must be an AST."
```

Comment:
I see, thats good! My only concern was that the current message was too generic and the whole pipe thing had no doc.

But I guess only with the docs now it can be helpful enough.

Reply:
To be clear, most users will read the docs of `Kernel.|>/2` and not `Macro.pipe`. `Macro.pipe` is the implementation detail. Are the docs of `Kernel.|>/2` clearer? Maybe we should redirect this function to the docs in `|>/2` instead?

Reply:
Oh, that makes sense, even with the name pipe I didn't saw it that way, because here its just appending expressions into an argument list on AST.

Reply:
I've removed this clause, and fixed some details on the doc recommended by @whatyouhide, we could add a link to pipe `Kernel.|>/2` docs at the end